### PR TITLE
fix problems in some loss operators

### PR DESCRIPTION
**Pull request by mosout**
- [x] bce_loss，nll_loss在reduction为none情况下tmp_buffer为nullptr无法获取mut_dptr
- [x] nll_loss 0-dim sbp推导
- [x] 优化算子的cuda实现

### CODE REVIEW

**Review by strint — APPROVED** [review]

File: oneflow/user/kernels/binary_cross_entropy_kernel.cu [hunk 1]
```diff
@@ -58,38 +68,42 @@ __global__ void ComputeBinaryCrossEntropyOut(int64_t elem_cnt, const half* input
...
+  const float eps = 1e-12;
   CUDA_1D_KERNEL_LOOP(i, elem_cnt) {
     float input_val = __half2float(input[i]);
     float target_val = __half2float(target[i]);
     float dy_val = __half2float(reduction_type == ReductionType::kNone ? dy[i] : *dy);
-
-    dx[i] =
-        __float2half(dy_val * (input_val - target_val) / max((1.0 - input_val) * input_val, 1e-12));
-    if (weight != nullptr) { dx[i] = __hmul(dx[i], weight[i]); }
+    float dx_val = __float2half(dy_val * (input_val - target_val)
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_kernel.cu:100]: 这里是不是应该是
```cpp
half dx_val = xxxxxx; 
```
因为你已经调用了 `__float2half`

File: oneflow/user/kernels/binary_cross_entropy_kernel.cu [hunk 2]
```diff
@@ -156,7 +169,7 @@ class BinaryCrossEntropyGradKernel final : public user_op::OpKernel {
         ctx->has_input("weight", 0) ? ctx->Tensor4ArgNameAndIndex("weight", 0)->dptr<T>() : nullptr;
     ComputeBinaryCrossEntropyGradOut<<<BlocksNum4ThreadsNum(elem_cnt), kCudaThreadsNumPerBlock, 0,
                                        ctx->device_ctx()->cuda_stream()>>>(
-        elem_cnt, input, target, dy, dx, weight, reduction);
+        elem_cnt, 1.0 / elem_cnt, input, target, dy, dx, weight, reduction);
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_kernel.cu:172]: 这里是不是要加个`static_cast<float>(1.0 / elem_cnt)`

不过函数参数这里指定了是float，理论上不加也可以

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu [hunk 3]
```diff
@@ -52,21 +57,24 @@ __global__ void ComputeBinaryCrossEntropyWithLogitsOut(int64_t elem_cnt, const h
...
+  const float one_val = 1.0;
   CUDA_1D_KERNEL_LOOP(i, elem_cnt) {
-    float input_val = __half2float(input[i]);
-    float target_val = __half2float(target[i]);
-    float max_val = -input_val < 0 ? 0 : -input_val;
-
+    const float input_val = __half2float(input[i]);
+    const float target_val = __half2float(target[i]);
+    const float max_val = -input_val < zero_val ? zero_val : -input_val;
+    float out_val;
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:66]: 这里应该是 `half out_val`

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu [hunk 4]
```diff
@@ -87,53 +95,60 @@ __device__ __forceinline__ float CalSigmoid(const float x) {
 }
 
 template<typename T>
-__global__ void ComputeBinaryCrossEntropyWithLogitsGradOut(int64_t elem_cnt, const T* input,
-                                                           const T* target, const T* dy, T* dx,
-                                                           const T* weight,
+__global__ void ComputeBinaryCrossEntropyWithLogitsGradOut(int64_t elem_cnt, float one_div_elem_cnt,
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:98]: `one_div_elem_cnt`有点chinglish

在layernorm kernel里，1.0/norm_size被命名为`inv_norm_size` https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/user/kernels/layer_norm_gpu_kernel.cu#L215

我还想了个 `elem_cnt_reciprocal`，你看看哪个好点

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu [hunk 5]
```diff
@@ -87,53 +95,60 @@ __device__ __forceinline__ float CalSigmoid(const float x) {
...
+      dx_val = __float2half((input_sigmoid - target_val) * dy_val);
     } else {
-      dx[i] =
-          __float2half(dy_val
-                       * ((__half2float(pos_weight_processed[i]) + 1 - target_val) * input_sigmoid
-                          - __half2float(pos_weight_processed[i])));
+      float pos_weight_processed_val = __half2float(pos_weight_processed[i]);
+      dx_val = __float2half(
+          dy_val
+          * ((pos_weight_processed_val + static_cast<float>(1) - target_val) * input_sigmoid
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:143]: `static_cast<float>(1` -> `static_cast<float>(1.0)`

不改也行

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu [hunk 6]
```diff
@@ -239,7 +254,7 @@ class BinaryCrossEntropyWithLogitsGradKernel final : public user_op::OpKernel {
     ComputeBinaryCrossEntropyWithLogitsGradOut<<<BlocksNum4ThreadsNum(elem_cnt),
                                                  kCudaThreadsNumPerBlock, 0,
                                                  ctx->device_ctx()->cuda_stream()>>>(
-        elem_cnt, input, target, dy, dx, weight, pos_weight_processed, reduction);
+        elem_cnt, 1.0 / elem_cnt, input, target, dy, dx, weight, pos_weight_processed, reduction);
```
Comment (MARD1NO) [inline oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:257]: 同上，这里应该也要个static_cast？

File: oneflow/user/kernels/kl_div_kernel.cu [hunk 7]
```diff
@@ -27,11 +27,15 @@ template<typename T>
...
+      const T target_val = target[i];
+      out[i] = exp(target_val) * (target_val - input[i]);
+    }
   } else {
     CUDA_1D_KERNEL_LOOP(i, elem_cnt) {
-      const auto out_val = target[i] * (SafeLog(target[i]) - input[i]);
-      out[i] = target[i] > 0 ? out_val : static_cast<T>(0);
+      const T target_val = target[i];
+      const auto out_val = target_val * (SafeLog(target_val) - input[i]);
+      out[i] = target_val > 0 ? out_val : static_cast<T>(0);
```
Comment (MARD1NO) [inline oneflow/user/kernels/kl_div_kernel.cu:38]: target_val 应该和 `static_cast<T>(0)` 比较

可以和你底下的half版本做法类似，单独提出来

File: oneflow/user/kernels/kl_div_kernel.cu [hunk 8]
```diff
@@ -107,8 +118,8 @@ class KLDivGradKernel : public SimpleLossGradKernel<DeviceType::kGPU, T, KLDivGr
                   const T* target, const T* dy, T* dx, const ReductionType reduction) const {
     const bool log_target = ctx->Attr<bool>("log_target");
     ComputeKLDivGradOut<<<BlocksNum4ThreadsNum(elem_cnt), kCudaThreadsNumPerBlock, 0,
-                          ctx->device_ctx()->cuda_stream()>>>(elem_cnt, input, target, dy, dx,
-                                                              reduction, log_target);
+                          ctx->device_ctx()->cuda_stream()>>>(
+        elem_cnt, 1.0 / elem_cnt, input, target, dy, dx, reduction, log_target);
```
Comment (MARD1NO) [inline oneflow/user/kernels/kl_div_kernel.cu:122]: 同上 1.0/elem_cnt也加个static_cast

File: oneflow/user/kernels/loss_kernel_util.cu [hunk 9]
```diff
@@ -32,8 +32,9 @@ __global__ void ApplyLossReductionImpl(int64_t elem_cnt, const T* tmp_out, T* ou
   __syncthreads();
   T block_sum = BlockReduce(cub_reduce_tmp_storage).Reduce(thread_sum, cub::Sum());
   if (threadIdx.x == 0) {
-    *out = block_sum;
-    if (is_reduce_mean) { *out /= elem_cnt; }
+    T out_val = block_sum;
+    if (is_reduce_mean) { out_val /= elem_cnt; }
```
Comment (MARD1NO) [inline oneflow/user/kernels/loss_kernel_util.cu:36]: 不要做除法，可以现在cpu算好，当作一个常数传到核函数这里

File: oneflow/user/kernels/loss_kernel_util.cu [hunk 10]
```diff
@@ -50,8 +51,9 @@ __global__ void ApplyLossReductionImpl<half>(int64_t elem_cnt, const half* tmp_o
   __syncthreads();
   half block_sum = BlockReduce(cub_reduce_tmp_storage).Reduce(thread_sum, cub::Sum());
   if (threadIdx.x == 0) {
-    *out = block_sum;
-    if (is_reduce_mean) { *out = __float2half(__half2float(*out) / elem_cnt); }
+    half out_val = block_sum;
+    if (is_reduce_mean) { out_val = __float2half(__half2float(*out) / elem_cnt); }
```
Comment (MARD1NO) [inline oneflow/user/kernels/loss_kernel_util.cu:55]: 同上

File: oneflow/user/kernels/nll_kernel.cu [hunk 11]
```diff
@@ -105,16 +112,18 @@ __global__ RETURN_VOID_IF_HALF ComputeNllOutReduce(const int64_t num_instances,
                                                    const K* target, T* out, const T* weight,
                                                    T* total_weight, bool is_reduce_mean) {
 #if __CUDA_ARCH__ >= 530 || !defined(__CUDA_ARCH__)
+  const T zero_val = __float2half(0.0);
```
Comment (MARD1NO) [inline oneflow/user/kernels/nll_kernel.cu:115]: `const T` -> `const half` ？

然后我感觉传参这里是不是可以显式指定为half了

File: oneflow/user/kernels/nll_kernel.cu [hunk 12]
```diff
@@ -200,6 +209,7 @@ class NllKernel final : public user_op::OpKernel {
     T* total_weight = total_weight_blob->mut_dptr<T>();
     const T* weight =
         ctx->has_input("weight", 0) ? ctx->Tensor4ArgNameAndIndex("weight", 0)->dptr<T>() : nullptr;
+    Memset<DeviceType::kGPU>(ctx->device_ctx(), total_weight, 0, GetCudaAlignedSize(sizeof(T)));
```
Comment (MARD1NO) [inline oneflow/user/kernels/nll_kernel.cu:212]: 这里不应该用`GetCudaAlignedSize`，你可以到 `oneflow/core/common/util.h`下对照公式代入计算下，结果并不是你想要的 单个T 的size，这样memset是会越界的

直接 `sizeof(T)` 应该就能满足你的要求了

File: oneflow/user/kernels/binary_cross_entropy_kernel.cu (hunk 1 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_kernel.cu:100]: 对

File: oneflow/user/kernels/binary_cross_entropy_kernel.cu (hunk 2 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_kernel.cu:172]: 这个地方用了一个隐式的类型转换  我显式一下吧

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu (hunk 3 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:66]: 对

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu (hunk 4 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:98]: 用inv_elem_cnt吧 简洁

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu (hunk 5 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:143]: 嗯嗯

File: oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu (hunk 6 above)
Comment (mosout) [reply oneflow/user/kernels/binary_cross_entropy_with_logits_kernel.cu:257]: 对

File: oneflow/user/kernels/kl_div_kernel.cu (hunk 7 above)
Comment (mosout) [reply oneflow/user/kernels/kl_div_kernel.cu:38]: 好的

File: oneflow/user/kernels/kl_div_kernel.cu (hunk 8 above)
Comment (mosout) [reply oneflow/user/kernels/kl_div_kernel.cu:122]: 好的

File: oneflow/user/kernels/loss_kernel_util.cu (hunk 9 above)
Comment (mosout) [reply oneflow/user/kernels/loss_kernel_util.cu:36]: 好

File: oneflow/user/kernels/loss_kernel_util.cu (hunk 10 above)
Comment (mosout) [reply oneflow/user/kernels/loss_kernel_util.cu:55]: 好

File: oneflow/user/kernels/nll_kernel.cu (hunk 11 above)
Comment (mosout) [reply oneflow/user/kernels/nll_kernel.cu:115]: 这个我感觉无所谓这个T肯定是half

File: oneflow/user/kernels/nll_kernel.cu (hunk 12 above)
Comment (mosout) [reply oneflow/user/kernels/nll_kernel.cu:212]: 好

**Comment by strint** [comment]
这个距离合并还差什么不，有个社区的小伙伴比较需要这个完成他们cpu 下的模型并行 demo

**Comment by mosout** [comment]
> 这个距离合并还差什么不，有个社区的小伙伴比较需要这个完成他们cpu 下的模型并行 demo

等我把上面review的意见改完应该就可以合并了

File: oneflow/user/kernels/smooth_l1_loss_kernel.cu [hunk 13]
```diff
@@ -59,47 +59,50 @@ __global__ void ComputeSmoothL1Out(int64_t elem_cnt, const half* input, const ha
...
+__global__ void ComputeSmoothL1GradOut(int64_t elem_cnt, float inv_elem_cnt, const T* input,
+                                       const T* target, const T* dy, T* dx,
+                                       const ReductionType reduction_type, const float beta) {
   CUDA_1D_KERNEL_LOOP(i, elem_cnt) {
     const T diff = input[i] - target[i];
     const T abs_diff = abs(diff);
+    T dx_val;
     if (abs_diff < beta) {
-      dx[i] = diff / beta;
+      dx_val = diff / beta;
```
Comment (MARD1NO) [inline oneflow/user/kernels/smooth_l1_loss_kernel.cu:70]: 这个也有修改的空间，可以给核函数多传一个 `inv_beta`

**Review by MARD1NO — APPROVED** [review]

File: oneflow/user/kernels/smooth_l1_loss_kernel.cu (hunk 13 above)
Comment (mosout) [reply oneflow/user/kernels/smooth_l1_loss_kernel.cu:70]: 好的
